framework/src/main/java/org/checkerframework/framework/util/ComplexQualifierHierarchy.java

```diff
@@ -0,0 +1,309 @@
+     */
+    protected Map<QualifierKind, AnnotationMirror> createBottoms() {
+        Map<QualifierKind, AnnotationMirror> bottomsMap = new TreeMap<>();
+        for (QualifierKind kind : qualifierKindHierarchy.getBottoms()) {
+            bottomsMap.put(kind, AnnotationBuilder.fromClass(elements, kind.getAnnotationClass()));
+        }
+        return bottomsMap;
+    }
+
+    /**
+     * Returns the qualifier kind for the given annotation.
+     *
+     * @param anno annotation mirror
+     * @return the qualifier kind for the given annotation
+     */
+    protected QualifierKind getQualifierKind(AnnotationMirror anno) {
+        String name = AnnotationUtils.annotationName(anno);
+        return getQualifierKind(name);
+    }
+
+    /**
+     * Returns the qualifier kind for the annotation with the fully qualified name {@code name}.
+     *
+     * @param name fully qualified annotation name
+     * @return the qualifier kind for the annotation with {@code name}
+     */
+    protected QualifierKind getQualifierKind(String name) {
+        QualifierKind kind = qualifierKindHierarchy.getQualifierKind(name);
+        if (kind == null) {
+            throw new BugInCF("QualifierKind not in hierarchy: %s", name);
+        }
+        return kind;
+    }
+
+    @Override
+    public Set<? extends AnnotationMirror> getTopAnnotations() {
+        return tops;
+    }
+
+    @Override
+    public AnnotationMirror getTopAnnotation(AnnotationMirror start) {
+        QualifierKind kind = getQualifierKind(start);
+        return topsMap.get(kind.getTop());
+    }
+
+    @Override
+    public Set<? extends AnnotationMirror> getBottomAnnotations() {
+        return bottoms;
+    }
+
+    @Override
+    public @Nullable AnnotationMirror getPolymorphicAnnotation(AnnotationMirror start) {
+        QualifierKind polyKind = getQualifierKind(start).getPolymorphic();
+        if (polyKind == null) {
+            return null;
+        }
+        AnnotationMirror poly = kindToAnnotationMirror.get(polyKind);
+        if (poly == null) {
+            throw new TypeSystemError(
+                    "Poly %s has an element. Override ComplexQualifierHierarchy#getPolymorphicAnnotation.",
+                    polyKind);
+        }
+        return poly;
+    }
+
+    @Override
+    public boolean isPolymorphicQualifier(AnnotationMirror qualifier) {
+        return getQualifierKind(qualifier).isPoly();
+    }
+
+    @Override
+    public AnnotationMirror getBottomAnnotation(AnnotationMirror start) {
+        QualifierKind kind = getQualifierKind(start);
+        return bottomsMap.get(kind.getBottom());
+    }
+
+    @Override
+    public boolean isSubtype(AnnotationMirror subAnno, AnnotationMirror superAnno) {
+        QualifierKind subKind = getQualifierKind(subAnno);
+        QualifierKind superKind = getQualifierKind(superAnno);
+        if (subKind.isSubtype(superKind)) {
+            if (superKind.hasElements() && subKind.hasElements()) {
```

Comment:
I expected this test to be `||` rather than `&&`.  There could be hierarchies like

```
    @Top
   /    \
@B(1)   @B(2)
```

or

```
@B(1)    @B(2)
  |        |
 @C       @D
```

Reply:
I agree that such a hierarchy could exist.  (I don't know of any.)  The current implementation of ComplexQualifierHierarchy doesn't support such a hierarchy.  If we change the implementation, then all the subclasses would have to expect those cases and add boilerplate: 
```
if (!superKind.hasElements() || !subKind.hasElements()) {
    return true;
}
```

I could add another subclass of QualifierHierarchy that allows hierarchies like the one you mentioned.  But what should it be called? I guess in general renaming `ComplexQualifierHierarchy` to `QualifierHierarchyWithElements` isn't exactly correct because it can't represent the all hierarchies with qualifiers that have elements.

I'll continue to think on this.